### PR TITLE
[CDEC-330] Remove cycle in `Pos/Core/Block` modules

### PR DESCRIPTION
**Pull request by ruhatch**
## Description

This is the first stage of clean up of `Pos.Core.Block` modules. It:

- Moves the `Blockchain` instances for `GenesisBlockchain` and `MainBlockchain` into `Pos.Core.Block.Union.Types`
- Removes the `data family` definition of `BlockHeader`
- Moves `HeaderHash` and `HasHeaderHash` into `Pos.Core.Block.Union.Types`
- A bunch of other small changes to remove any module cycles

The next stage would:

- Break up the `Pos.Core.Class` module, pushing overloaded accessors back to their original types
- Clean up the `Instances` and `Lens` files in the `Pos.Core.Block` set of modules
- Rename the `Pos.Core.Block.Union.Types` to `Pos.Core.Block`, replacing the existing reexport file

This would leave us with only four files:

- `Pos.Core.Block.Blockchain` containing `Blockchain` class and `GenericBlock(Header)` definitions
- `Pos.Core.Block.Genesis` containing most types and instances for `Blockchain GenesisBlockchain`
- `Pos.Core.Block.Main` containing most types and instances for `Blockchain MainBlockchain`
- `Pos.Core.Block` import the other three and assembling the `Blockchain` instances

## Linked issue

[CDEC-330](https://iohk.myjetbrains.com/youtrack/issue/CDEC-330)

## Type of change

- [~] 🐞 Bug fix (non-breaking change which fixes an issue)
- [~] 🛠 New feature (non-breaking change which adds functionality)
- [~] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [~] 🔨 New or improved tests for existing code
- [~] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [x] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist

- [~] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: core/src/Pos/Core/Block/Union/Types.hs [hunk 1]
```diff
@@ -17,34 +36,158 @@ module Pos.Core.Block.Union.Types
...
+    { _msHeaderHash  :: !HeaderHash  -- ^ Hash of previous header
+                                     --    in the chain
+    , _msBodyProof   :: !MainProof
+    , _msSlot        :: !SlotId
+    , _msChainDiff   :: !ChainDifficulty
+    , _msExtraHeader :: !MainExtraHeaderData
+    } deriving Generic
+
+deriving instance Show MainToSign
+deriving instance Eq MainToSign
```
[the remark this answers is not in the document]
Comment (avieth) [reply core/src/Pos/Core/Block/Union/Types.hs]: I prefer standalone deriving always. Do we have any style guide dogma about this?

File: core/src/Pos/Core/Block/Genesis/Instances.hs [hunk 2]
```diff
@@ -15,19 +15,20 @@ import           Formatting (bprint, build, int, sformat, stext, (%))
...
+import           Pos.Core.Block.Union.Types (BlockHeader (..), GenesisBlock, GenesisBlockHeader,
+                                             HeaderHash, blockHeaderHash)
 import           Pos.Core.Class (HasDifficulty (..), HasEpochIndex (..), HasEpochOrSlot (..),
                                  HasHeaderHash (..), IsGenesisHeader, IsHeader)
-import           Pos.Core.Common (HeaderHash, slotLeadersF)
+import           Pos.Core.Common (slotLeadersF)
 import           Pos.Core.Slotting.Types (EpochOrSlot (..))
 import           Pos.Crypto (hashHexF)
 
+instance NFData GenesisBlock
```
[the remark this answers is not in the document]
Comment (ruhatch) [reply core/src/Pos/Core/Block/Genesis/Instances.hs]: That will be part of the next PR on these files - this whole `Instances` file should disappear!

File: core/src/Pos/Core/Block/Union/Types.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (parsonsmatt) [reply core/src/Pos/Core/Block/Union/Types.hs]: With new GHC features IMO we should be using datatype-attached deriving clauses almost always, so we can have mandatory `DerivingStrategies` :smile: 

Standalone deriving clauses that don't add anything extra to the context feel smelly to me.

File: core/src/Pos/Core/Block/Union/Instances.hs [hunk 3]
```diff
@@ -53,6 +56,13 @@ getBlockHeader = \case
 blockHeader :: Getter Block BlockHeader
 blockHeader = to getBlockHeader
 
+-- | Representation of 'Block' passed to a component.
+data ComponentBlock payload =
+    ComponentBlockGenesis (Some IsGenesisHeader)
+    | ComponentBlockMain
+       { bcmHeader  :: !(Some IsMainHeader)
+       , bcmPayload :: !payload }
```
[the remark this answers is not in the document]
Comment (ruhatch) [reply core/src/Pos/Core/Block/Union/Instances.hs]: Yup - @erikd is getting golden/property tests in place before we remove them to make sure we can safely change the types

File: core/src/Pos/Core/Block/Union/Types.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ruhatch) [reply core/src/Pos/Core/Block/Union/Types.hs]: Good question - could it be to do with the deriving strategy used?

File: core/src/Pos/Core/Block/Union/Instances.hs [hunk 4]
```diff
@@ -53,6 +56,13 @@ getBlockHeader = \case
 blockHeader :: Getter Block BlockHeader
 blockHeader = to getBlockHeader
 
+-- | Representation of 'Block' passed to a component.
```
[the remark this answers is not in the document]
Comment (ruhatch) [reply core/src/Pos/Core/Block/Union/Instances.hs]: Very good question!

File: core/src/Pos/Core/Block/Union/Instances.hs [hunk 5]
```diff
@@ -53,6 +56,13 @@ getBlockHeader = \case
...
+data ComponentBlock payload =
+    ComponentBlockGenesis (Some IsGenesisHeader)
+    | ComponentBlockMain
+       { bcmHeader  :: !(Some IsMainHeader)
```
[the remark this answers is not in the document]
Comment (ruhatch) [reply core/src/Pos/Core/Block/Union/Instances.hs]: Yeah, I didn't add anything in this patch. There's a whole bunch of unnecessarily abstract stuff, but it's not especially consistent anyway.

**Comment by ruhatch** [comment]
@avieth No worries. Yeah I didn't want things to get out of control. I'd definitely appreciate your thoughts on what would be safe to change in another round of refactorings for this.

**Comment by ruhatch** [comment]
@avieth Just audited the usage of the `Blockchain` class and it's not used in more than a couple of places. I'll get going on removing it right now! I'll still put it into a separate PR though I think.

**Pull request merged**